test/mpi/init/session.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpi.h"
+#include <stdio.h>
+#include <assert.h>
+#include "mpitest.h"
+
+/* This is example 10.8 from MPI Standard 4.0 */
+
+int errs = 0;
+
+int library_foo_test(void);
+void library_foo_init(void);
+void library_foo_finalize(void);
+
+int main(int argc, char *argv[])
+{
+#ifdef MULT_INIT
+    int rank, size, provided;
+    int num_repeat = 1;
+
+    if (argc > 1) {
+        num_repeat = atoi(argv[1]);
+        /* basic sanity check */
+        assert(num_repeat > 0 && num_repeat < 100);
+    }
+
+    MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    for (int i = 0; i < num_repeat; i++) {
+        library_foo_test();
+    }
+    MPI_Finalize();
+#else
+    int rank = library_foo_test();
+#endif
+    if (rank == 0 && errs == 0) {
+        printf("No Errors\n");
+    }
+    return MTestReturnValue(errs);
+}
+
+static MPI_Session lib_shandle = MPI_SESSION_NULL;
+static MPI_Comm lib_comm = MPI_COMM_NULL;
+
+int library_foo_test(void)
+{
+    int rank, size;
+
+    library_foo_init();
+
+    MPI_Comm_size(lib_comm, &size);
+    MPI_Comm_rank(lib_comm, &rank);
+
+    int sum;
+    MPI_Reduce(&rank, &sum, 1, MPI_INT, MPI_SUM, 0, lib_comm);
+    if (rank == 0) {
+        if (sum != (size - 1) * size / 2) {
+            printf("MPI_Reduce: expect %d, got %d\n", (size - 1) * size / 2, sum);
+            errs++;
+        }
+    }
+
+    library_foo_finalize();
+
+    return rank;
+}
+
+void library_foo_init(void)
+{
+    int rc, flag;
+    int ret = 0;
+    const char pset_name[] = "mpi://WORLD";
+    const char mt_key[] = "mpi_thread_support_level";
+    const char mt_value[] = "MPI_THREAD_MULTIPLE";
+    char out_value[100];        /* large enough */
+
+    MPI_Group wgroup = MPI_GROUP_NULL;
+    MPI_Info sinfo = MPI_INFO_NULL;
+    MPI_Info tinfo = MPI_INFO_NULL;
+    MPI_Info_create(&sinfo);
+    MPI_Info_set(sinfo, mt_key, mt_value);
+    rc = MPI_Session_init(sinfo, MPI_ERRORS_RETURN, &lib_shandle);
+    if (rc != MPI_SUCCESS) {
+        errs++;
+        goto fn_exit;
+    }
+
+    /* check we got thread support level foo library needs */
+    rc = MPI_Session_get_info(lib_shandle, &tinfo);
+    if (rc != MPI_SUCCESS) {
+        errs++;
+        goto fn_exit;
+    }
+
+    MPI_Info_get(tinfo, mt_key, sizeof(out_value), out_value, &flag);
+    if (flag != 1) {
```

Comment:
> Standard defines the value of flag returned by MPI_Info_get to be true or false. Thus, better to use if(!flag) rather than if(flag != 1), as the test may be used by different MPI implementations.

will fix

Reply:
done